src/v/datalake/schema_avro.cc

```diff
@@ -0,0 +1,316 @@
+    }
+
+    state.type_hierarchy.push_back(node);
+    auto deferred_pop = ss::defer(
+      [&state] { state.type_hierarchy.pop_back(); });
+
+    switch (node->type()) {
+    case avro::AVRO_STRING:
+        if (node->logicalType().type() == avro::LogicalType::UUID) {
+            return iceberg::uuid_type{};
+        }
+
+        return iceberg::string_type{};
+    case avro::AVRO_BYTES:
+        if (node->logicalType().type() == avro::LogicalType::DECIMAL) {
+            return create_decimal(node);
+        }
+        return iceberg::binary_type{};
+    case avro::AVRO_INT: {
+        auto lt = node->logicalType().type();
+        if (lt == avro::LogicalType::DATE) {
+            return iceberg::date_type{};
+        }
+        if (lt == avro::LogicalType::TIME_MILLIS) {
+            return iceberg::time_type{};
+        }
+        return iceberg::int_type{};
+    }
+    case avro::AVRO_LONG: {
+        auto lt = node->logicalType().type();
+        if (lt == avro::LogicalType::TIME_MICROS) {
+            return iceberg::time_type{};
+        }
+        if (
+          lt == avro::LogicalType::TIMESTAMP_MILLIS
+          || lt == avro::LogicalType::TIMESTAMP_MICROS) {
+            return iceberg::timestamp_type{};
+        }
+        return iceberg::long_type{};
+    }
+    case avro::AVRO_FLOAT:
+        return iceberg::float_type{};
+    case avro::AVRO_DOUBLE:
+        return iceberg::double_type{};
+    case avro::AVRO_BOOL:
+        return iceberg::boolean_type{};
+    case avro::AVRO_NULL:
+        return std::nullopt;
+    case avro::AVRO_RECORD: {
+        auto struct_result = struct_from_avro_record(node, state);
+        if (struct_result.has_error()) {
+            return struct_result.error();
+        }
+        return std::make_optional<iceberg::field_type>(
+          std::move(struct_result.value()));
+    }
+    case avro::AVRO_ENUM:
+        return iceberg::int_type{};
+    case avro::AVRO_ARRAY: {
+        if (node->leaves() != 1) {
+            return schema_conversion_exception(fmt::format(
+              "Invalid number of leaves {} in AVRO_ARRAY, expected to have "
+              "only 1 leaf",
+              node->leaves()));
+        }
+        auto array_id = state.next_field_id();
+        auto element_type = node->leafAt(0);
+        auto field_res = field_type_from_avro(element_type, state);
+        if (field_res.has_error()) {
+            return field_res.error();
+        }
+        if (!field_res.value().has_value()) {
+            return schema_conversion_exception(
+              fmt::format("Unsupported type of AVRO_NULL as an array element"));
+        }
+        return iceberg::list_type::create(
+          array_id,
+          iceberg::field_required::yes,
+          std::move(*field_res.value()));
+    }
+    case avro::AVRO_MAP: {
+        // Avro Map is represented as an Iceberg List with key value pairs as we
+        // can not guarantee the uniqueness of keys in AVRO_MAP
```

Comment:
What do we do for protobuf here? Also what do libraries for avro do? Do they take the latest key? That is the proto semantics.

I ask because I think this is could be a mismatch of user expectations

Reply:
good point, i will change it to map to `iceberg::map`. In avro codegen map is generated as: 
```
  case avro::AVRO_MAP:
            return "std::map<std::string, " + cppTypeOf(n->leafAt(1)) + " >";
```